### ISSUE TITLE
quad_float performance

### ISSUE THREAD

**Issue by ckormanyos** [BoostGSoC21/multiprecision#67]
So with the first drafts of quad-float available, this issue touches on the performance of operators and functions for `cpp_quad_float` compared with, for example, `cpp_bin_float` at similar precision.

**Comment by ckormanyos** [comment]
Check this out, just did a very, very, very preliminary multiplication timing featuring `cpp_quad_float<double>` versus `cpp_bin_float` with a similar digit count.

I found (with faster/smaller time being better):

```
quad_time: 0.793
ctrl_time: 2.102
result_is_ok: true
```

I don't think quad or double will put dec or bin out of business, but there is promise here for a range of applications that handle the wily nature of these composite type(s).

I will follow this up with much more analysis moving forward.

Cc: @sinandredemption and @cosurgi and @jzmaddock

**Comment by sinandredemption** [comment]
> I will follow this up with much more analysis moving forward.

Yes, please do. It will be interesting to see the performance comparison to `mpf_float` or `gmp_float`, which are the currently the the fastest types Boost provides if I'm not mistaken.

**Comment by ckormanyos** [comment]
> interesting to see the performance comparison to `mpf_float` or `gmp_float`, which are the currently the the fastest types Boost provides

Actually, this is billed as expected to a good race because the GMP/MPIR types suffer from expensive allocation at this low-to-moderate digit count less than or similar to 30...100 decimal digits. Let's stabilize quad-float and along the way I'll throw in more performance stuff.

@jzmaddock has a performance suite that we will/should proabalby activate as soon as our backend types are a bit more stable. i'll dust this off and start taking a look at how it's playing in the performance suite.

**Comment by ckormanyos** [comment]
After #101 I got the following timing. All-in-all for a first cut, performance looks promising.

Certain laces in particular need to be handled/optimized in more detail:
- all forms of division
- construct-from-string
- mixed arithmetic operations with `long` `long`


| operation                        |  cpp_bin_float  |   cpp_quad_fp_backend   |
| -------------------------------- | --------------- | ----------------------- |
|  +                               |   0.0331187     |  0.0246576              |
|  -                               |   0.0344958     |  0.0267205              |
|  *                               |   0.0452308     |  0.0378971              |
|  /                               |   0.548875      |  0.474183               |
|  str                             |   0.0018916     |  0.0301673              |
|  +(int)                          |   0.0327899     |  0.0099693              |
|  -(int)                          |   0.0340673     |  0.0099627              |
|  *(int)                          |   0.0186423     |  0.0150899              |
|  /(int)                          |   0.0843419     |  0.212649               |
|  construct                       |   0.006122      |  0.0004477              |
|  construct(unsigned)             |   0.0040559     |  0.0012088              |
|  construct(unsigned long long)   |   0.0038952     |  0.0513031              |
|  +(unsigned long long)           |   0.04225       |  0.0518809              |
|  -(unsigned long long)           |   0.0488046     |  0.0531701              |
|  *(unsigned long long)           |   0.0290197     |  0.064957               |
|  /(unsigned long long)           |   0.082562      |  0.235138               |
|  +=(unsigned long long)          |   0.0438786     |  0.0508877              |
|  -=(unsigned long long)          |   0.0439513     |  0.0502721              |
|  *=(unsigned long long)          |   0.0220493     |  0.0632934              |
|  /=(unsigned long long)          |   0.100602      |  0.236258               |

Cc: @sinandredemption and @cosurgi

**Comment by jzmaddock** [comment]
It might be interesting to compare to the original double_double implementation as well?

BTW it's possible to use mpfr without memory allocation, for example `number<mpfr_float_backend<32, allocate_stack> >` would be a broadly comparable type I would guess.

**Comment by ckormanyos** [comment]
> might be interesting to compare to the original double_double implementation as well?

Yes, understood thanks John. Indeed, @jzmaddock, there is a lot to do in general to clean up our double-float implementation. This includes characterizing its perf also.

> possible to use mpfr without memory allocation

That's cool. I must admit, I fumbled around a bit with `performance_test.cpp`. At first I compared to `cpp_bin_float` but had used too many digits in bin-float. At the moment, The numbres just above in the table are getting better, but honestly quad-double still needs quite a bit of optimization effort.

I'll keep on this with your suggestions and report as we go along...

**Comment by sinandredemption** [comment]
It just crossed my mind, we can maybe use [GodBolt](https://www.godbolt.org/) to compare the assembly code generated with `std::tie` and `std::get<>` vs old-school assignments.

**Comment by ckormanyos** [comment]
Yes. I can also show you how to generate assembly listings from either msvc or GCC as well. There are a couple of good options to figure out what's going on under the hood...

**Comment by ckormanyos** [comment]
> compare the assembly code

Try to start with a reduced program. I might suggest to begin with multiplication. I can work out a reduced example if you like which brings the assembly code more or less clearly to light...

**Comment by sinandredemption** [comment]
I ran a quick benchmark vs QD package (IEEE-complaint additions operations). Here are the results (100000000 operations):

| Operator | cpp_quad_float | qd_real |
| ---------- | ---------------- | -------- |
| +             | 9.083s              | 10.693  |
| *              | 22.126s            | 21.394s |

So seems like the `std::get<>`/`std::tie` overhead isn't too much of an issue, and we might need to try optimization methods on the algorithmic level.

CC @ckormanyos @cosurgi

**Comment by ckormanyos** [comment]
> the `std::get<>`/`std::tie` overhead isn't too much of an issue, and we might need to try optimization methods on the algorithmic level.

That is a really good data point to have, Fahad. Good work @sinandredemption.

I have no problem putting this issue on hold for the moment. Thoughts?

**Comment by cosurgi** [comment]
> I have no problem putting this issue on hold for the moment. Thoughts?

It is okay for me.

**Comment by ckormanyos** [comment]
I am closing this issue for the context of GSoC '21. It was never part of this project to re-invent such algorithms, rather to make them Boost-ified, with modern support, portable and well-tested.

If we need to re-open this issue at a later time or in another form, we can then easily do so at such a later time, should one arise...

Cc: @sinandredemption and @cosurgi
